plugin/core/transports.py

```diff
@@ -48,26 +49,33 @@ def on_stderr_message(self, message: str) -> None:
 
 class AbstractProcessor(Generic[T]):
 
-    def write_data(self, writer: IO[bytes], data: T) -> None:
+    def write_data(self, writer: IO[bytes], data: T, is_node_ipc: bool) -> None:
         raise NotImplementedError()
 
-    def read_data(self, reader: IO[bytes]) -> Optional[T]:
+    def read_data(self, reader: IO[bytes], is_node_ipc: bool) -> Optional[T]:
         raise NotImplementedError()
 
 
 class JsonRpcProcessor(AbstractProcessor[Dict[str, Any]]):
 
-    def write_data(self, writer: IO[bytes], data: Dict[str, Any]) -> None:
+    def write_data(self, writer: IO[bytes], data: Dict[str, Any], is_node_ipc: bool) -> None:
         body = self._encode(data)
-        writer.writelines(("Content-Length: {}\r\n\r\n".format(len(body)).encode('ascii'), body))
+        if not is_node_ipc:
+            writer.writelines(("Content-Length: {}\r\n\r\n".format(len(body)).encode('ascii'), body))
+        else:
+            writer.write(body + b"\n")
+
+    def read_data(self, reader: IO[bytes], is_node_ipc: bool) -> Optional[Dict[str, Any]]:
+        if not is_node_ipc:
+            headers = http.client.parse_headers(reader)  # type: ignore
+            try:
+                body = reader.read(int(headers.get("Content-Length")))
+            except TypeError:
+                # Expected error on process stopping. Stop the read loop.
+                raise StopLoopError()
+        else:
+            body = reader.readline()
```

Comment:
Since the logic is pretty much separate for node_ipc, it sounds like there should be a separate processor for that case.

Reply:
I agree! The current name is a bit confusing, since it's an LSP processor, and JSON RPC only describes the structure of the payload. How would you name it? And the Node.js IPC equivalent?

Reply:
I would suggest to not extende the AbstracProcessor class with a is_node_ipc flag argument 
and to not modify the JsonRpcProcessor class.

Instead create a separate class that extends the AbstractProcessor class, called NodeIpcProcessor (feel free to name it better).

And instantiate that NodeIpcProcessor instead of JsonRpcProcessor if an explicit property flag in the client configuration field exist.

That way we avoid if statements that were added because of the is_node_ipc flag argument.

Reply:
The defining property of this node IPC protocol seems to be that messages are separated by a newline character.  Contrast to JSON-RPC, which uses HTTP headers to separate messages. I'm surprised it's that simple. I think having it called NodeIpcProcessor makes sense. All the logic of the current NodeIpcIO class can go in there.